compiler/next/include/chpl/uast/Conditional.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_CONDITIONAL_H
+#define CHPL_UAST_CONDITIONAL_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/BlockStyle.h"
+#include "chpl/uast/ControlFlow.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents a conditional. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      const flag = false;
+      if flag then writeln('a'); else writeln('b');
+
+  \endrst
+
+ */
+class Conditional final : public ControlFlow {
+ private:
+  Conditional(ASTList children, int8_t conditionChildNum,
```

Comment:
I think this mimics the AST layout used by `LLVM` and helps to avoid extra indirection when talking about the stuff in the body of a conditional/loop. Another argument is that it also avoids associating semantic actions related to e.g. scoping with the `Block` node or codifying an implicit expectation of that in the compiler.

Reply:
Yeah, we could always store a `Block` for the then/else body. Likewise we could have the body of a `for` loop always be a `Block`. We have so far opted not to do that - instead favoring a flatter AST design. From a philosophical point of view, it is debatable whether or not `if a then b;` and `if a then { b; }` both contain a block, syntactically. It's my view that by hiding this detail we reduce potential for confusion when working with the AST nodes and the AST class documentation will still make it obvious how to e.g. access the statements in the `then`  part.

Anyway the statement forms of these really do introduce a new scope for the then/else code. We might revisit this choice if it proves to be awkward to do scope resolution without having such Blocks.

Reply:
I think @stonea also asked about the constructor arguments. At least I interpreted it being partly "why not have the constructor accept separate child lists for "then" and "else" branches. The way I understand that aspect is that we want constructors to be simple, and any necessary preprocessing to be done in build* functions.